pkg/scheduler/scheduler.go

```diff
@@ -140,14 +140,15 @@ func (s *Scheduler) schedule(ctx context.Context) {
 			continue
 		}
 		cq := snapshot.ClusterQueues[e.ClusterQueue]
-		if e.assignment.Borrows() && cq.Cohort != nil && usedCohorts.Has(cq.Cohort.Name) {
-			e.status = skipped
-			e.inadmissibleMsg = "workloads in the cohort that don't require borrowing were prioritized and admitted first"
-			continue
-		}
-		// Even if there was a failure, we shouldn't admit other workloads to this
-		// cohort.
 		if cq.Cohort != nil {
+			e.assignment.Borrows()
+			if usedCohorts.Has(cq.Cohort.Name) && (e.assignment.Borrows() || cq.Cohort.HasBorrowingQueues()) {
```

Comment:
can you explain why this new check solves the bug in the PR description and what other alternatives you considered and why they didn't work?

I'm wondering whether this check is excessive.

IIUC, the over-admission would happen in the second scheduling cycle:
- one workload was already preempted
- two workloads try to use the space, the first one fits.
- The second one is not aware that the first one was just scheduled. Could we just call `snapshot.AddWorkload` when we call `AssumeWorkload`?

And this also makes me think that there is a bug even without preemption, in an scenario very similar to the integration test you have below. Disable preemption and just mark one of the beta workloads as finished after you have already created the workloads for alpha and gamma.

Reply:
Please add a comment to this check to summarize why we are doing it.

Reply:
Please add the comment

Reply:
done